crates/apollo-compiler/examples/extract_directives_used_by_query.rs

```diff
@@ -0,0 +1,101 @@
+use apollo_compiler::{
+    hir::{Directive, Field},
+    ApolloCompiler, FileId, HirDatabase,
+};
+
+fn get_directives_used_in_query(compiler: &ApolloCompiler, query_id: &FileId) -> Vec<Directive> {
+    // seed the stack with top-level fields
+    let mut stack: Vec<Field> =
+        compiler
+            .db
+            .operations(*query_id)
+            .iter()
+            .fold(vec![], |mut acc, operation_definition| {
+                acc.extend(operation_definition.selection_set().fields().to_vec());
+                acc
+            });
+
+    let mut directives = vec![];
+
+    // depth first search for nested fields with directives
+    while !stack.is_empty() {
+        if let Some(field) = stack.pop() {
```

Comment:
Here `stack.is_empty()` and `stack.pop()` are basically doing the same check, so it can be simplified to
```rust
while let Some(field) = stack.pop() {
```

I'll apply that change before merging!